### PR TITLE
Non static statistics: Round 1

### LINKED ISSUE #4499: StatisticsLevel.Verbose causes serialization exceptions
- Orleans 2.0

**config:**
``` csharp
client = new ClientBuilder()
    .Configure<ClusterOptions>(options =>
    {
        options.ClusterId = clusterId;
        options.ServiceId = ServiceId.ToString();
    })
    .Configure<ClientStatisticsOptions>(options =>
    {
        options.CollectionLevel = StatisticsLevel.Verbose;
    })
    .UseZooKeeperClustering((ZooKeeperGatewayListProviderOptions options) =>
    {
        options.ConnectionString = zookeeper;
    })
    .ConfigureApplicationParts(parts => parts.AddApplicationPart(typeof(ISampleGrain).Assembly))
    .ConfigureLogging(logging=>logging.AddConsole())
    .Build();
```

**log:**
> warn: Messaging.GatewayConnection/GatewayClientSender_gwy.tcp://10.0.1.18:10300/262055360[100159]                               
      Unexpected error serializing message to gateway gwy.tcp://10.0.1.18:10300/262055360.                                      
System.NullReferenceException: Object reference not set to an instance of an object.                                            
   at Orleans.Serialization.SerializationManager.SerializeMessageHeaders(HeadersContainer headers, SerializationContext context)
   at Orleans.Runtime.Message.Serialize(SerializationManager serializationManager, Int32& headerLengthOut, Int32& bodyLengthOut)
   at Orleans.Messaging.OutgoingMessageSender.Process(Message msg)                                                              
warn: Orleans.OutsideRuntimeClient[104202]                                                                                      
      Refresh the GrainTypeResolver failed. WIll be retried after                                                               
Orleans.Runtime.OrleansMessageRejectionException: Unexpected error serializing message to gateway gwy.tcp://10.0.1.18:10300/262055360. System.NullReferenceException: Object reference not set to an instance of an object.
    at Orleans.Serialization.SerializationManager.SerializeMessageHeaders(HeadersContainer headers, SerializationContext context)
    at Orleans.Runtime.Message.Serialize(SerializationManager serializationManager, Int32& headerLengthOut, Int32& bodyLengthOut)
    at Orleans.Messaging.OutgoingMessageSender.Process(Message msg)                                                              
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()                                                            
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)                            
    at Orleans.Runtime.OutgoingCallInvoker.<Invoke>d__21.MoveNext()

### PR DESCRIPTION
**Pull request by ReubenBond**
Note that this is based upon #4509
Fixes #4499

Submitting for CI testing, but it passed in VSO already.

This starts the process of removing statics from the statistics system. It doesn't complete the process, but it is already enough to fix initialization order problems like we see in #4499.

### CODE REVIEW

File: src/Orleans.Core/Runtime/CallbackData.cs
Comment (ReubenBond) [reply src/Orleans.Core/Runtime/CallbackData.cs]: I went ahead and removed StatisticsCollector

File: src/Orleans.Core/Serialization/SerializationManager.cs
Comment (ReubenBond) [reply src/Orleans.Core/Serialization/SerializationManager.cs]: I did this in a later commit

File: src/Orleans.Core/Statistics/StatisticsCollector.cs
Comment (xiazen) [inline src/Orleans.Core/Statistics/StatisticsCollector.cs]: this change looks good to me. the only thing I would suggest is since we splitted `StatisticsCollector ` and now each StatisticGroupe handle the bool on should the system collect their metrics or not in their own logic. This is consistent with the direction we are going with. But we also lost this single place we can look up on what does `StatisticsLevel ` really mean. Set it to a specific level would enable which metrics.  So I suggest put this info in `statisticsOptions.CollectionLevel ` xml comment

File: src/Orleans.Core/Runtime/OutsideRuntimeClient.cs
Comment (xiazen) [inline src/Orleans.Core/Runtime/OutsideRuntimeClient.cs]: this variable shouldn't be on the OutsideRuntimeClient level, it should be with the StatisticGroup which collects `ThreadTimeTrackingStats`

File: src/Orleans.Core/Messaging/ClientMessageCenter.cs
Comment (xiazen) [inline src/Orleans.Core/Messaging/ClientMessageCenter.cs]: this variable shouldn't be part of ClientMessagesCenter, and set value when ClientMessagesCenter was constructed. It should be with the StatisticGroup which collects QueueStats

Comment (xiazen) [reply src/Orleans.Core/Messaging/ClientMessageCenter.cs]: so maybe something like:
```
class QueueTrackingStatistic
{
public static bool ShouldCollect(CollectionLevel level) => level>= StatisticLevel.Verbose2;
public bool CollectQueueStats {get;set;}
```
And code here will be: `if(QueueTrackingStatistic.ShouldCollect(statisticOptions.CollectionLevel)) queueTracking = ...`

File: src/Orleans.Core/Messaging/ClientMessageCenter.cs
Comment (xiazen) [inline src/Orleans.Core/Messaging/ClientMessageCenter.cs]: and here will just be queueTracking?.OnSTartExecution().

File: src/Orleans.Core/Messaging/ClientMessageCenter.cs
Comment (xiazen) [reply src/Orleans.Core/Messaging/ClientMessageCenter.cs]: logic on whether or not collect a set of statistics should stay with the StatisticGroup.

File: src/Orleans.Core/Threading/ThreadPoolExecutor.cs
Comment (xiazen) [inline src/Orleans.Core/Threading/ThreadPoolExecutor.cs]: I think this logic should be in `ThreadTrackingStatistic`. The logic on whether to collect a set of statistics should be in the statisticGroup itself

File: src/Orleans.Core/Threading/ThreadPoolThread.cs
Comment (xiazen) [inline src/Orleans.Core/Threading/ThreadPoolThread.cs]: same here

File: src/Orleans.Runtime/Messaging/InboundMessageQueue.cs
Comment (xiazen) [inline src/Orleans.Runtime/Messaging/InboundMessageQueue.cs]: same here

File: src/Orleans.Core/Statistics/StatisticsCollector.cs
Comment (ReubenBond) [reply src/Orleans.Core/Statistics/StatisticsCollector.cs]: Sounds good to me, I'll come up with a better solution to keep some of the semantics localized

**Comment by ReubenBond** [comment]
Feedback addressed

File: src/Orleans.Core/Messaging/ClientMessageCenter.cs
Comment (xiazen) [inline src/Orleans.Core/Messaging/ClientMessageCenter.cs]: I was more on move CollectQueueStats() method to `QueueTrackingStatistic` itself, since you did that in other StatisticGroup classes. So to keep consistent, we should move `bool CollectXXXStats()` into the `XXXStatisticGroup` class.

But this is fine, it is more like how it was before, as in whether to collect a Stats is in the same place of `CollectionLevel` settings. Only problem is that the pattern is not consistent over all StatisticGroup.  For example, `ApplicationRequestsStatisticsGroup ` has the `CollectXXXStats` method within itself. 

I will leave it to you to decide do you want to keep it consistent or not.

**Review by xiazen — APPROVED** [review]

**Comment by xiazen** [comment]
yo I talked with Sergey, he is OK with the breaking change in #4509 . And since this PR is build on top of that and don't have other breaking changes, how about we rebase this one, merge this one and close #4509 after?

**Comment by ReubenBond** [comment]
@dotnet-bot test this please

**Comment by ReubenBond** [comment]
@dotnet-bot test bvt

**Pull request merged**
